pkg/logql/log/labels.go

```diff
@@ -1,19 +1,137 @@
 package log
 
-// Labels is the type that is passed across multiple stages.
-// I expect this type to become more and more complex over time as we optimize it.
-type Labels map[string]string
+import (
+	"sort"
 
-func (l Labels) Has(key string) bool {
-	_, ok := l[key]
-	return ok
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// LabelsBuilder is the same as labels.Builder but tailored for this package.
+type LabelsBuilder struct {
+	base labels.Labels
+	del  []string
+	add  []labels.Label
+
+	err string
+}
+
+// NewLabelsBuilder creates a new labels builder.
+func NewLabelsBuilder() *LabelsBuilder {
+	return &LabelsBuilder{
+		del: make([]string, 0, 5),
+		add: make([]labels.Label, 0, 5),
+	}
+}
+
+// Reset clears all current state for the builder.
+func (b *LabelsBuilder) Reset(base labels.Labels) {
+	b.base = base
+	b.del = b.del[:0]
+	b.add = b.add[:0]
+	b.err = ""
 }
 
-func (l Labels) SetError(err string) {
-	l[ErrorLabel] = err
+// SetErr sets the error label.
+func (b *LabelsBuilder) SetErr(err string) *LabelsBuilder {
+	b.err = err
+	return b
 }
 
-func (l Labels) HasError() bool {
-	_, ok := l[ErrorLabel]
-	return ok
+// GetErr return the current error label value.
+func (b *LabelsBuilder) GetErr() string {
+	return b.err
+}
+
+// HasErr tells if the error label has been set.
+func (b *LabelsBuilder) HasErr() bool {
+	return b.err != ""
+}
+
+// Base returns the base labels unmodified
+func (b *LabelsBuilder) Base() labels.Labels {
+	return b.base
+}
+
+func (b *LabelsBuilder) Get(key string) (string, bool) {
+	for _, a := range b.add {
+		if a.Name == key {
+			return a.Value, true
+		}
+	}
+Outer:
+	for _, l := range b.base {
+		if l.Name == key {
+			for _, d := range b.del {
+				if d == key {
+					continue Outer
```

Comment:
I think the continue is unnecessary. If it's found in the deletion group you can return `"", false`, no?

Reply:
yep you're right changing it and adding a test to cover it.